File: src/youtube_buddy/protagonist.py
from __future__ import annotations

import hashlib
import re
from dataclasses import dataclass
from datetime import datetime, timedelta

from youtube_buddy.config import Settings
from youtube_buddy.corrections import CorrectionStore


_NAME_PATTERN = re.compile(r"\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,2})\b")


@dataclass(slots=True)
class ProtagonistCandidate:
    name: str
    confidence: float
    source_mix: list[str]


@dataclass(slots=True)
class ProtagonistState:
    video_id: str
    ready: bool
    candidate: ProtagonistCandidate | None
    started_at: datetime


class ProtagonistTracker:
    def __init__(self, settings: Settings, corrections: CorrectionStore | None = None):
        self.settings = settings
        self.corrections = corrections or CorrectionStore()
        self._state: ProtagonistState | None = None

    def _video_id_from_frame(self, screenshot_png: bytes) -> str:
        return hashlib.sha1(screenshot_png).hexdigest()[:16]

    def ensure_state(self, screenshot_png: bytes, now: datetime) -> ProtagonistState:
        video_id = self._video_id_from_frame(screenshot_png)
        if self._state is None or self._state.video_id != video_id:
            self._state = ProtagonistState(
                video_id=video_id,
                ready=False,
                candidate=None,
                started_at=now,
            )
        return self._state

    def update(self, screenshot_png: bytes, transcript_hint: str, now: datetime) -> ProtagonistState:
        state = self.ensure_state(screenshot_png, now)
        candidate = self._infer_candidate(state.video_id, transcript_hint)
        preload = self.settings.protagonist.preload_window_seconds
        age = now - state.started_at
        ready = candidate is not None and (
            candidate.confidence >= self.settings.protagonist.min_confidence
            or age >= timedelta(seconds=preload)
        )
        state.ready = ready
        state.candidate = candidate
        return state

    def apply_correction(self, corrected_name: str) -> None:
        if not self._state:
            return
        key = self._correction_key(self._state.video_id, corrected_name)
        self.corrections.boost(key)
        self._state.candidate = ProtagonistCandidate(
            name=corrected_name,
            confidence=0.95,
            source_mix=["user_correction"],
        )
        self._state.ready = True

    def _correction_key(self, video_id: str, name: str) -> str:
        return f"{video_id}:{name.lower()}"

    def _infer_candidate(self, video_id: str, transcript_hint: str) -> ProtagonistCandidate | None:
        names = [m.group(1) for m in _NAME_PATTERN.finditer(transcript_hint)]
        if not names:
            return None
        winner = max(names, key=len)
        key = self._correction_key(video_id, winner)
        base_confidence = min(0.8, 0.35 + len(winner) / 24)
        confidence = min(0.99, base_confidence + self.corrections.score(key))
        return ProtagonistCandidate(
            name=winner,
            confidence=confidence,
            source_mix=["transcript"],
        )

    @property
    def state(self) -> ProtagonistState | None:
        return self._state

File: ios/Models/Protagonist.swift
import Foundation

struct ProtagonistCandidate {
    let name: String
    var confidence: Double
    var sourceMix: [String]
}

struct ProtagonistState {
    let videoId: String
    var ready: Bool
    var candidate: ProtagonistCandidate?
    let startedAt: Date
}

final class ProtagonistTracker {
    private let settings: AppSettings
    private let corrections: CorrectionStore
    private var currentState: ProtagonistState?

    private static let namePattern = try! NSRegularExpression(
        pattern: #"\b([A-Z][a-z]+(?:\s+[A-Z][a-z]+){0,2})\b"#
    )

    init(settings: AppSettings, corrections: CorrectionStore) {
        self.settings = settings
        self.corrections = corrections
    }

    var state: ProtagonistState? { currentState }

    /// Derive a pseudo video-id from page title (since we don't have frame hashing on iOS).
    private func videoId(from metadata: VideoMetadata?) -> String {
        guard let meta = metadata, !meta.videoId.isEmpty else {
            return "unknown-\(UUID().uuidString.prefix(8))"
        }
        return meta.videoId
    }

    func ensureState(metadata: VideoMetadata?, now: Date) -> ProtagonistState {
        let vid = videoId(from: metadata)
        if let existing = currentState, existing.videoId == vid {
            return existing
        }
        let fresh = ProtagonistState(videoId: vid, ready: false, candidate: nil, startedAt: now)
        currentState = fresh
        return fresh
    }

    @discardableResult
    func update(metadata: VideoMetadata?, transcriptHint: String, now: Date) -> ProtagonistState {
        var state = ensureState(metadata: metadata, now: now)

        // Build hint text from metadata + transcript
        var hints = transcriptHint
        if let meta = metadata {
            hints += " " + meta.title + " " + meta.channelName
        }

        let candidate = inferCandidate(videoId: state.videoId, text: hints)
        let preload = TimeInterval(settings.protagonist.preloadWindowSeconds)
        let age = now.timeIntervalSince(state.startedAt)

        let ready: Bool
        if let c = candidate {
            ready = c.confidence >= settings.protagonist.minConfidence || age >= preload
        } else {
            ready = age >= preload  // After preload window, proceed even without candidate
        }

        state.ready = ready
        state.candidate = candidate
        currentState = state
        return state
    }

    func applyCorrection(_ name: String) {
        guard var state = currentState else { return }
        let key = correctionKey(videoId: state.videoId, name: name)
        corrections.boost(key: key)
        state.candidate = ProtagonistCandidate(
            name: name,
            confidence: 0.95,
            sourceMix: ["user_correction"]
        )
        state.ready = true
        currentState = state
    }

    // MARK: - Private

    private func correctionKey(videoId: String, name: String) -> String {
        "\(videoId):\(name.lowercased())"
    }

    private func inferCandidate(videoId: String, text: String) -> ProtagonistCandidate? {
        let range = NSRange(text.startIndex..., in: text)
        let matches = Self.namePattern.matches(in: text, range: range)
        let names = matches.compactMap { match -> String? in
            guard let r = Range(match.range(at: 1), in: text) else { return nil }
            return String(text[r])
        }
        guard !names.isEmpty else { return nil }

        let winner = names.max(by: { $0.count < $1.count })!
        let key = correctionKey(videoId: videoId, name: winner)
        let baseConfidence = min(0.8, 0.35 + Double(winner.count) / 24.0)
        let boosted = min(0.99, baseConfidence + corrections.score(for: key))

        return ProtagonistCandidate(
            name: winner,
            confidence: boosted,
            sourceMix: ["transcript"]
        )
    }
}
